server/admin_management/api/app/services/access_control_privilege_service.py

```diff
@@ -157,3 +176,45 @@ def grant_privilege(
             )
 
         return access_control_privilege_return
+
+
+class ForestClientValidator:
+    """
+    Purpose: More validations on inputs (other than basic validations) and
+             business rules validations (if any).
+    Cautious: Do not instantiate the class for more than one time per request.
+              It calls Forest Client API remotely if needs to.
+    """
```

Comment:
In access_control_privilege_service, it has a **loop** and instantiates this "**ForestClientValidator**" several times depending on number of the forest_client_number in the user requests.

I wonder how slow it is for this looping, since Forest Client API itself response is already slow enough....plus now each loop it needs to do a new extra instantiation for this  "ForestClientValidator" class and "ForestClientService" class as well for the setup for the whole classes...

In normal usage, a "service"(validator is also a service) is actually designed to be used as a **singleton** per request (which currently our python code does not enforce it yet).

Maybe you like to adjust the validator class a little so that you can move this validator instantiation outside of the loop at line 84 `validator = ForestClientValidator(forest_client_number)` and only pass forest_client_number to a validate function of the class.
Or you can have this class take an array of forest_client_number (including 1 item) and have the looping logic inside the class.

Reply:
Hi Ian, thanks! that's a good point, I forgot it's a loop now, I'll try to make some adjustment!

Reply:
Hi Ian, I updated the validator do you mind check it again? thanks!!